### PR TITLE
Add Winforms-compatible asyncio Proactor.

### PR DESCRIPTION
**Pull request by freakboy3742**
Integrates the Winforms event handling lifecycle with the asyncio event loop.

This is done by instantiating a Proactor, but customizing the Proactor so that `run_forever()` is split into two parts; `setup_run_forever()` and `tick()`. The `tick()` component is what historically would have been run as a tight "while True" loop; in this version, the call to tick is scheduled to run every 5ms using the winforms event infrastructure.

It also involves some refactoring of the winforms libs - they were starting to get unwieldly, and introducing the Proactor was a good opportunity to modularize that code.

Huge thanks for @hawkowl for the many, many pointers on getting this going.

### CODE REVIEW

File: src/winforms/toga_winforms/libs/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from .winforms import (   # noqa: E401
+    WinForms, Single,
```
Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.WinForms' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Single' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Convert' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.WinDateTime' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Uri' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.WinIcon' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.WinImage' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.WinFont' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.ContentAlignment' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Size' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Point' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.TextAlignment' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.HorizontalTextAlignment' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.FontFamily' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.FontStyle' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.win_font_family' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Text' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Color' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.Bitmap' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.add_handler' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.user32' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.win_version' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/__init__.py]: At column 1: (F401) '.winforms.shcore' imported but unused

File: src/winforms/toga_winforms/widgets/numberinput.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import sys
-from toga_winforms.libs import LEFT, RIGHT, CENTER
-from toga_winforms.libs import WinForms, Convert, HorizontalTextAlignment
+
+from toga.constants import LEFT, RIGHT, CENTER
```
Comment (brutusthebee) [inline src/winforms/toga_winforms/widgets/numberinput.py]: At column 1: (F401) 'toga.constants.LEFT' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/widgets/numberinput.py]: At column 1: (F401) 'toga.constants.RIGHT' imported but unused

Comment (brutusthebee) [inline src/winforms/toga_winforms/widgets/numberinput.py]: At column 1: (F401) 'toga.constants.CENTER' imported but unused

**Comment by freakboy3742** [comment]
This is a work in progress; there's one more improvement to come. The current implementation relies on the existence of a main window, and ties the delayed execution to the form for that window. @hawkowl suggested an additional change to use ApplicationContext() as the main runnable, which would enable us to decouple the "application" lifecycle from the main form.

**Comment by freakboy3742** [comment]
The underlying bug preventing a full transition to an AppContext version of this patch can be found here: https://github.com/pythonnet/pythonnet/issues/965

File: src/winforms/toga_winforms/app.py [hunk 3]
```diff
@@ -11,15 +15,20 @@ def on_close(self):
         pass
 
 
+
 class App:
```
Comment (brutusthebee) [inline src/winforms/toga_winforms/app.py]: At column 1: (E303) too many blank lines (3)

File: src/winforms/toga_winforms/libs/proactor.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            controlling the lifecycle of the app.
+        """
+        # Remember the application context.
+        self.app_context = app_context
+
+        # Register a custom user window message.
+        self.msg_id = user32.RegisterWindowMessageA("Python asyncio tick")
+        # Add a message filter to listen for the asyncio tick message
+        msg_filter = AsyncIOTickMessageFilter(self, self.msg_id)
+        # FIXME: Actually install the message filter.
```
Comment (brutusthebee) [inline src/winforms/toga_winforms/libs/proactor.py]: At column 9: (F841) local variable 'msg_filter' is assigned to but never used

File: src/winforms/toga_winforms/app.py (hunk 3 above)
Comment (brutusthebee) [inline src/winforms/toga_winforms/app.py]: At column 1: (E303) too many blank lines (3)

**Comment by freakboy3742** [comment]
I've merged this because it doesn't appear that there's a simple fix for the Python.net binding, and an asyncio implementation that works *at all* is better than what we have right now. 

I've opened #750 to track the future work that can be done to improve this implementation.

**Pull request merged**